File: src/classical/linear_regression/linear_regression.py
import numpy as np
from classical.classical_model import ClassicalModel

class LinearRegression(ClassicalModel):
    """
    A simple implementation of Linear Regression using Ordinary Least Squares (OLS)
    """
    
    def __init__(self) -> None:
        """Initialize the model parameters"""
        self.params = {'beta': None}
    
    def _validate_input(self, array1: np.ndarray, array2: np.ndarray = None) -> None:
        """
        Validate the input numpy arrays

        Args:
            array (np.ndarray): Input numpy array
        """
        if not isinstance(array1, np.ndarray):
            raise TypeError("Array 1 must be a numpy array")
        
        if len(array1.shape) != 2:
            raise ValueError("Array 1 must be a 2D array")
        
        if array1.size == 0:
            raise ValueError("Array 1 must not be empty")
        
        if array2 is not None:
            if not isinstance(array2, np.ndarray):
                raise TypeError("Array 2 must be a numpy array")
        
            if len(array2.shape) != 2:
                raise ValueError("Array 2 must be a 2D array")
            
            if array2.size == 0:
                raise ValueError("Array 2 must not be empty")
            
            # Check if the number of rows in X and Y are equal
            if array1.shape[0] != array2.shape[0]:
                raise ValueError("The number of rows in X and Y must be equal")

    def fit(self, X: np.ndarray, Y: np.ndarray) -> None:
        """
        Fit the Linear Regression model to the training data

        Args:
            X (np.ndarray): Feature matrix of shape (n_samples, n_features)
            Y (np.ndarray): Target matrix of shape (n_samples, 1)
        """
        # Validate the input arrays
        self._validate_input(X,Y)
        
        # Evaluate the model parameters using OLS
        self.params['beta'] = np.linalg.pinv(X.T @ X) @ X.T @ Y

    def predict(self, X: np.ndarray) -> np.ndarray:
        """
        Predict target values using the learned model

        Args:
            X (np.ndarray): Feature matrix of shape (n_samples, n_features)

        Returns:
            np.ndarray: Predicted values of shape (n_samples, n_targets)
        """
        if not self.params:
            raise ValueError("Fit the model before making predictions")
        
        self._validate_input(X)
        
        if X.shape[1] != self.params['beta'].shape[0]:
            raise ValueError("The number of columns in X must be equal to the number of features in the model")
        
        return X @ self.params['beta']

    def cost(self, Y_pred: np.ndarray, Y: np.ndarray) -> float:
        """
        Compute the Mean Squared Error cost

        Args:
            Y_pred (np.ndarray): Predicted target values
            Y (np.ndarray): True target values

        Returns:
            float: Mean Squared Error
        """
        self._validate_input(Y_pred, Y)
        SCALE_FACTOR = 0.5 / Y.shape[0]
        return SCALE_FACTOR * np.linalg.norm(Y_pred - Y)**2

    def get_params(self) -> dict:
        """
        Retrieve model parameters

        Returns:
            dict: Model parameters {'beta': np.ndarray}
        """
        return self.params

File: src/classical/linear_regression/c_backend/linear_regression.c
/**
 * @file linear_regression.c
 * @brief Linear regression model functions for fitting, prediction, and cost calculation.
 *
 * This file contains the implementation of a basic linear regression model. It includes functions for fitting
 * the model using ordinary least squares (OLS), predicting the output based on input features, and calculating 
 * the cost (mean squared error).
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "linear_regression.h"
#include "c_mathematical_functions/pseudoinverse.h"
#include "c_loss_functions/loss_functions.h"
#include "c_memory_functions/memory_functions.h"
#include <gsl/gsl_matrix.h>

/**
 * @brief Fit the linear regression model to the given data.
 *
 * This function computes the parameters (Beta) of the linear regression model by solving the normal equation.
 * It uses the formula Beta = (X^T * X)^(-1) * X^T * Y, where X is the input matrix, Y is the target matrix, 
 * and Beta are the model parameters (weights). The pseudoinverse of (X^T * X) is computed to handle the inversion.
 *
 * @param[in] X The input matrix of size (num_samples x num_input_features).
 * @param[in] Y The target/output matrix of size (num_samples x num_output_features).
 * @param[out] Beta The weights of the model, of size (num_input_features x num_output_features).
 * @param[in] num_samples The number of training samples.
 * @param[in] num_input_features The number of features in the input matrix X.
 * @param[in] num_output_features The number of output features in Y.
 */
void fit(const float *X, const float *Y, float *Beta, const int num_samples, const int num_input_features, const int num_output_features) {
    // Calculate the inner product of the input matrix and allocate memory for the result
    float *XTX;
    XTX = (float*) safeMalloc(num_input_features * num_input_features * sizeof(float));
    cblas_sgemm(
        CblasRowMajor, CblasTrans, CblasNoTrans,
        num_input_features, num_input_features, num_samples,
        1,
        X, num_input_features,
        X, num_input_features,
        0,
        XTX,num_input_features
        );
    
    // Take the pseudoinverse of the inner product and allocate memory for the result
    float *XTX_inverse;
    XTX_inverse = (float*) safeMalloc(num_input_features * num_input_features * sizeof(float));
    computePseudoinverse(XTX, XTX_inverse, num_input_features, num_input_features);

    // Free intermediate
    safeFree(XTX);

    // Multiply the pseudoinverse and the input matrix and allocate memory for the result
    float *XTX_inverse_XT;
    XTX_inverse_XT = (float*) safeMalloc(num_input_features * num_samples * sizeof(float));
    cblas_sgemm(
        CblasRowMajor, CblasNoTrans, CblasTrans,
        num_input_features, num_samples, num_input_features,
        1,
        XTX_inverse, num_input_features,
        X, num_input_features,
        0,
        XTX_inverse_XT, num_samples
        );
    
    // Free intermediate
    safeFree(XTX_inverse);

    // Multiply the running product with Y
    cblas_sgemm(
        CblasRowMajor, CblasNoTrans, CblasNoTrans,
        num_input_features, num_output_features, num_samples,
        1,
        XTX_inverse_XT, num_samples,
        Y, num_output_features,
        0,
        Beta, num_output_features
        );
    
    // Free intermediate
    safeFree(XTX_inverse_XT);
}

/**
 * @brief Predict the output based on the input features using the trained model.
 *
 * This function predicts the output using the linear regression model by multiplying the input matrix X
 * with the learned weights Beta. The result is stored in the Prediction matrix.
 *
 * @param[in] X The input matrix of size (num_samples x num_input_features).
 * @param[in] Beta The model weights, of size (num_input_features x num_output_features).
 * @param[out] Prediction The predicted output, of size (num_samples x num_output_features).
 * @param[in] num_samples The number of input samples.
 * @param[in] num_input_features The number of input features in the model.
 * @param[in] num_output_features The number of output features.
 */
void predict(const float *X, const float *Beta, float *Prediction, const int num_samples, const int num_input_features, const int num_output_features) {
    // Multiply the input matrix and weights
    cblas_sgemm(
        CblasRowMajor, CblasNoTrans, CblasNoTrans,
        num_samples, num_output_features, num_input_features,
        1,
        X, num_input_features,
        Beta, num_output_features,
        0,
        Prediction, num_output_features
        );
}

/**
 * @brief Calculate the cost (mean squared error) of the model.
 *
 * This function computes the mean squared error (MSE) between the predicted values and the actual values.
 * It is used to evaluate the performance of the model after training.
 *
 * @param[in] Y_pred The predicted output matrix, of size (num_samples x num_output_features).
 * @param[in] Y The actual target/output matrix, of size (num_samples x num_output_features).
 * @param[in] num_samples The number of samples.
 * @param[in] num_output_features The number of output features.
 *
 * @return The mean squared error between the predicted and actual outputs.
 */
float cost(const float *Y_pred, const float *Y, const int num_samples, const int num_output_features) {
    // Return the mean squared error
    return meanSquaredError(Y_pred, Y, num_samples, num_output_features);
}